draft-ietf-tls-esni.md

```diff
@@ -444,10 +445,33 @@ values, ClientHelloInner MUST also contain:
  - an "echo_nonce" extension
  - TLS padding {{!RFC7685}}
 
-Padding SHOULD be P = L - D bytes, where
-
-- L = ECHOConfig.maximum_name_length, rounded up to the nearest multiple of 16
-- D = len(dns_name), where dns_name is the DNS name in the ClientHelloInner "server_name" extension
+An encoded ClientHelloInner can be around 300 octets long, but variations in
+the length of the ciphertext version of that could defeat the entire purpose of
+ECHO, if for example those expose the chosen server_name field in the
+ClientHelloInner.  The padding ClientHello extension, if well used, can ensure
+that length information does not expose what is contained in the
+ClientHelloInner. Clients MUST add padding to the ClientHelloInner to try meet
+this requirement.
+
+Given that new extensions may be defined in future that contain sensitive
+length fields, we cannot describe all the ways in which length information
+could expose sensitive content. Implementers ought therefore be aware that they
+might have to change their padding scheme as the set of supported extensions
+changes. In addition to padding the ClientHelloInner, clients and servers MUST
+both pad all other handshake messages that have sensitive length fields. For
+example, if the client proposes ALPN values in the ClientHelloInner, the
+server-selected value will be returned in an EncryptedExtension, so that
+handshake message needs to be padded with TLS record layer padding. 
+
+One approach that may be relative safe and simple is once ECHO is in use:
+
+- When constructing the ClientHelloInner, use padding bytes so that the length
+  of the plaintext erncoding of ClienHelloInner is greater than
+ECHOConfig.minimum_inner_length and is a multiple of 16
+- Both clients and servers pad all handshake messages so that plaintexts are a
+  multiple of 16 octets long.
+- Clients or servers MAY send additional padding bytes. This MUST NOT be
+  treated as an error.
```

Comment:
What about the converse?  Is it an error to send something that is not padded to the minimum length?

Reply:
I think I'd say no, that's maybe better not treated as an error. But I could be convinced otherwise.